File: chat/internal/tui/rewind.go
package tui

import (
	"context"
	"fmt"
	"strconv"
	"strings"
)

// /rewind and /diff: checkpoints, rewind and the session diff (the web's
// rewind.ts; docs/claude-parity.md, item 11).

// rewindTargets are the chat's own user messages, oldest first, the ones
// a rewind can name (a subagent's prompt is not one).
func rewindTargets(c *Chat) []Entry {
	var out []Entry
	for _, e := range c.Conversation.Entries {
		if e.Role == "user" && e.ParentID == "" {
			out = append(out, e)
		}
	}
	return out
}

// rewindScope reads the scope word: code, conversation (conv) or both;
// both when empty.
func rewindScope(word string) (string, bool) {
	switch strings.ToLower(word) {
	case "", "both", "all":
		return "both", true
	case "code", "files":
		return "code", true
	case "conv", "conversation", "chat":
		return "conversation", true
	}
	return "", false
}

// rewind lists the messages (no argument) or asks to confirm a rewind to
// before message N with the scope given, then runs it.
func (a *App) rewind(ctx context.Context, c *Chat, arg string) {
	if c == nil {
		a.setNotice("no chat selected")
		return
	}
	targets := rewindTargets(c)
	if len(targets) == 0 {
		a.setNotice("no messages to rewind to")
		return
	}
	fields := strings.Fields(arg)
	if len(fields) == 0 {
		known := map[string]bool{}
		if list, err := a.Client.Checkpoints(ctx, c.ID); err == nil {
			for _, cp := range list {
				known[cp.ID] = true
			}
		}
		var b strings.Builder
		for i, e := range targets {
			mark := " "
			if known[e.ID] {
				mark = "•"
			}
			fmt.Fprintf(&b, "%2d %s %s\n", i+1, mark, truncate(excerptOf(e.Text), 70))
		}
		b.WriteString("/rewind N [code|conv|both] goes back to before message N (• has a checkpoint; code needs one)")
		a.setNotice(b.String())
		return
	}
	n, err := strconv.Atoi(fields[0])
	if err != nil || n < 1 || n > len(targets) {
		a.setNotice("/rewind N with N from /rewind")
		return
	}
	scope := ""
	if len(fields) > 1 {
		scope = fields[1]
	}
	what, ok := rewindScope(scope)
	if !ok || len(fields) > 2 {
		a.setNotice("/rewind N [code|conv|both]")
		return
	}
	if c.Running() {
		a.setNotice("stop the agent first (Esc)")
		return
	}
	target := targets[n-1]
	label := map[string]string{"code": "code", "conversation": "conversation", "both": "code and conversation"}[what]
	id := c.ID
	a.confirm = &confirmation{
		prompt: fmt.Sprintf("Rewind to before message %d %q (%s)? Type y and Enter to confirm; anything else cancels", n, truncate(excerptOf(target.Text), 60), label),
		run: func(ctx context.Context) {
			result, err := a.Client.Rewind(ctx, id, target.ID, what)
			if err != nil {
				a.setNotice(err.Error())
				return
			}
			a.diff = nil
			a.refreshState(ctx)
			a.setNotice(rewindNotice(result, n, label))
		},
	}
}

// undoHint is the line under a rewind marker while the rewind can still
// be undone: how, and what the workspace does.
func undoHint(e Entry) string {
	hint := "/undo-rewind puts the removed messages back (until the next turn)"
	if e.Rewind != nil && e.Rewind.What == "both" {
		if e.Rewind.Before != "" {
			hint += "; /undo-rewind code restores the workspace as it was before the rewind too"
		} else {
			hint += "; the workspace stays as it is"
		}
	}
	return hint
}

// undoRewind is /undo-rewind [code]: what the last conversation rewind
// removed goes back in place, after confirmation; `code` asks for the
// workspace as it was before the rewind too (a both-rewind's).
func (a *App) undoRewind(ctx context.Context, c *Chat, arg string) {
	if c == nil {
		a.setNotice("no chat selected")
		return
	}
	if c.UndoRewind == "" {
		a.setNotice("nothing to undo: no rewind since the last turn")
		return
	}
	code := false
	switch strings.ToLower(strings.TrimSpace(arg)) {
	case "":
	case "code", "files", "both", "all":
		code = true
	default:
		a.setNotice("/undo-rewind [code]")
		return
	}
	if c.Running() {
		a.setNotice("stop the agent first (Esc)")
		return
	}
	var marker Entry
	for _, e := range c.Conversation.Entries {
		if e.ID == c.UndoRewind {
			marker = e
			break
		}
	}
	if code && (marker.Rewind == nil || marker.Rewind.Before == "") {
		a.setNotice("no checkpoint of the workspace before the rewind was recorded; /undo-rewind restores the conversation and leaves the workspace as it is")
		return
	}
	scope := "the conversation"
	if code {
		scope = "the conversation and the workspace"
	}
	id, markerID := c.ID, c.UndoRewind
	a.confirm = &confirmation{
		prompt: fmt.Sprintf("Undo the rewind %q and put %s back? Type y and Enter to confirm; anything else cancels", truncate(excerptOf(marker.Text), 70), scope),
		run: func(ctx context.Context) {
			result, err := a.Client.UndoRewind(ctx, id, markerID, code)
			if err != nil {
				a.setNotice(err.Error())
				return
			}
			a.diff = nil
			a.refreshState(ctx)
			a.setNotice(undoNotice(result))
		},
	}
}

// undoNotice says what an undo did, as its transcript line does.
func undoNotice(r UndoResult) string {
	parts := []string{fmt.Sprintf("rewind undone: %d entries restored", r.Entries)}
	if r.Requeued > 0 {
		parts = append(parts, fmt.Sprintf("%d message(s) queued again and held", r.Requeued))
	}
	switch r.Session {
	case "cancelled":
		parts = append(parts, "the agent's session never saw the rewind")
	case "resumed":
		parts = append(parts, "the agent's session continues where it was")
	case "fresh":
		parts = append(parts, "the agent's session cannot take the messages back; the next message starts a new one with the conversation so far as context")
	}
	switch r.Code {
	case "restored":
		parts = append(parts, fmt.Sprintf("the workspace is back as it was before the rewind (%d file(s) restored, %d removed)", len(r.Restored), len(r.Removed)))
	case "kept":
		parts = append(parts, "the workspace stays as the rewind left it")
	}
	return strings.Join(parts, "; ")
}

// rewindNotice says what a rewind did, as the marker's second line does.
func rewindNotice(r RewindResult, n int, label string) string {
	parts := []string{fmt.Sprintf("rewound to before message %d (%s)", n, label)}
	if r.What != "conversation" {
		if len(r.Restored)+len(r.Removed) == 0 {
			parts = append(parts, "the workspace already matched the checkpoint")
		} else {
			parts = append(parts, fmt.Sprintf("%d file(s) restored, %d removed", len(r.Restored), len(r.Removed)))
		}
	}
	switch r.Conversation {
	case "pending":
		parts = append(parts, "the agent forgets the messages when its session resumes")
	case "fresh":
		parts = append(parts, "the agent's session could not rewind; the next message starts a new one with the conversation so far as context")
	}
	if r.Withdrawn > 0 {
		parts = append(parts, fmt.Sprintf("%d queued message(s) withdrawn", r.Withdrawn))
	}
	return strings.Join(parts, "; ")
}

// excerptOf is a message's first words on one line.
func excerptOf(text string) string {
	line := strings.Join(strings.Fields(text), " ")
	if line == "" {
		return "(attachments)"
	}
	return line
}

// showDiff toggles the session diff under the transcript: fetched now,
// shown folded per file until Tab expands; /diff again (or an argument
// of "close") hides it.
func (a *App) showDiff(ctx context.Context, c *Chat, arg string) {
	if c == nil {
		a.setNotice("no chat selected")
		return
	}
	if a.diff != nil || arg == "close" || arg == "off" {
		a.diff = nil
		a.setNotice("changes hidden")
		return
	}
	changes, err := a.Client.Diff(ctx, c.ID)
	if err != nil {
		a.setNotice(err.Error())
		return
	}
	a.diff = changes
	a.scroll = 0
	if len(changes.Files) == 0 {
		a.setNotice("no changes since this chat began")
		return
	}
	a.setNotice(fmt.Sprintf("%d changed file(s); Tab expands the hunks, /diff hides them", len(changes.Files)))
}

// RenderChanges lays out the session diff: a heading with the counts,
// then each file as one line (path, +added −removed), followed by its
// hunks when expanded.
func RenderChanges(d *WorkspaceChanges, width int, expanded bool) []string {
	added, removed := 0, 0
	for _, f := range d.Files {
		added += f.Added
		removed += f.Removed
	}
	head := fmt.Sprintf("Changes since this chat began: %d file(s), %s+%d%s %s−%d%s", len(d.Files), green, added, reset, red, removed, reset)
	if d.Truncated {
		head += dim + " (diff cut at 2 MiB)" + reset
	}
	out := wrap(bold+head+reset, width, "", "  ")
	if len(d.Files) == 0 {
		return append(out, dim+"  no changes"+reset)
	}
	hunks := splitDiff(d.Diff)
	for _, f := range d.Files {
		counts := fmt.Sprintf("%s+%d%s %s−%d%s", green, f.Added, reset, red, f.Removed, reset)
		if f.Binary {
			counts = dim + "binary" + reset
		}
		out = append(out, wrap(bold+sanitize(f.Path)+reset+"  "+counts, width, "  ", "    ")...)
		if !expanded {
			continue
		}
		text, ok := hunks[f.Path]
		if !ok {
			out = append(out, dim+"    │ (not in the diff)"+reset)
			continue
		}
		lines, _, _ := diffLines(text)
		out = append(out, renderBody(lines, width, true, false)...)
	}
	if !expanded {
		out = append(out, dim+"  Tab expands the hunks"+reset)
	}
	return out
}

// splitDiff cuts one git diff into its files by the path each `diff
// --git` line names (the new path).
func splitDiff(diff string) map[string]string {
	out := map[string]string{}
	current := ""
	for _, line := range strings.SplitAfter(diff, "\n") {
		if strings.HasPrefix(line, "diff --git ") {
			header := strings.TrimSuffix(line, "\n")
			current = strings.TrimPrefix(header, "diff --git ")
			if i := strings.Index(current, " b/"); i >= 0 {
				current = current[i+3:]
			}
		}
		if current != "" {
			out[current] += line
		}
	}
	return out
}

File: chat/web/src/rewind.ts
/* Checkpoints, rewind and the session diff: the pure parts. A checkpoint is
   the workspace as it was before one user message (the runner records it
   under the message's ID); a rewind goes back to before a message — its
   code, the conversation, or both — and the session diff is the workspace
   now against the chat's first checkpoint (or its last code rewind).

   A conversation (or both) rewind can be undone until the next turn
   starts or another rewind happens: the service keeps what it removed
   and `chat.undoRewind` names the marker whose Undo puts it back. The
   session follows as far as it can (a pending rewind is cancelled, a
   dropped thread resumed, a session that rewound starts fresh with the
   restored transcript as context); the workspace stays as the rewind
   left it unless the checkpoint the rewind recorded of it is restored
   (`entry.rewind.before`, a both-rewind's). */

import type { Chat, Entry } from "./types";

export type RewindWhat = "code" | "conversation" | "both";

export type Checkpoint = {
  id: string;
  chatID: string;
  commit: string;
  tree: string;
  store: string;
  changed: boolean;
  at: string;
};

export type RewindResult = {
  messageID: string;
  what: RewindWhat;
  restored?: string[];
  removed?: string[];
  conversation?: "rewound" | "pending" | "fresh" | "";
  /* How many queued messages a conversation rewind withdrew (queue.ts). */
  withdrawn?: number;
};

/* What undoing a rewind did (chats.UndoResult). */
export type UndoResult = {
  messageID: string;
  what: RewindWhat;
  entries: number;
  requeued?: number;
  session: "cancelled" | "resumed" | "fresh" | "";
  code?: "restored" | "kept" | "";
  restored?: string[];
  removed?: string[];
};

export type ChangedFile = {
  path: string;
  added: number;
  removed: number;
  binary?: boolean;
};

export type SessionChanges = {
  base: string;
  files: ChangedFile[];
  diff: string;
  truncated?: boolean;
};

/* One message the chat can rewind to: the user message, its number among
   them (from 1) and whether the runner holds a checkpoint for it. */
export type RewindTarget = {
  entry: Entry;
  number: number;
  checkpoint: boolean;
};

export const WHAT_LABELS: Record<RewindWhat, { label: string; hint: string }> =
  {
    both: {
      label: "Code and conversation",
      hint: "The workspace goes back to before the message and the agent forgets it and what followed",
    },
    conversation: {
      label: "Conversation only",
      hint: "The agent forgets the message and what followed; files stay as they are",
    },
    code: {
      label: "Code only",
      hint: "The workspace goes back to before the message; the transcript stays",
    },
  };

/* The user messages one can rewind to, oldest first. A subagent's prompt
   (a user entry under a task card) is not one. */
export function rewindTargets(
  entries: Entry[],
  checkpoints: Iterable<string> = [],
): RewindTarget[] {
  const known = new Set(checkpoints);
  const out: RewindTarget[] = [];
  for (const entry of entries) {
    if (entry.role !== "user" || entry.parentID) continue;
    out.push({
      entry,
      number: out.length + 1,
      checkpoint: known.has(entry.id),
    });
  }
  return out;
}

/* Whether a rewind can be asked for now: the chat is neither busy nor
   archived. The service checks the same. */
export function canRewind(chat: Pick<Chat, "status" | "archived">): boolean {
  return (
    !chat.archived &&
    chat.status !== "running" &&
    chat.status !== "queued" &&
    chat.status !== "stopping"
  );
}

/* Whether a scope can be picked for a target: code needs its checkpoint. */
export function whatAllowed(what: RewindWhat, target: RewindTarget): boolean {
  return what === "conversation" || target.checkpoint;
}

/* A message's first words for a list or a marker. */
export function excerpt(text: string, max = 80): string {
  const line = text.replace(/\s+/g, " ").trim();
  if (!line) return "(attachments)";
  const chars = Array.from(line);
  return chars.length > max ? chars.slice(0, max).join("") + "…" : line;
}

/* Esc pressed twice within `window` ms, Claude Code's rewind key. */
export function doubleEscape(last: number, now: number, window = 600): boolean {
  return last > 0 && now - last <= window;
}

export function changesSummary(files: ChangedFile[]): {
  files: number;
  added: number;
  removed: number;
} {
  let added = 0;
  let removed = 0;
  for (const f of files) {
    added += f.added;
    removed += f.removed;
  }
  return { files: files.length, added, removed };
}

/* One git diff split into its files, by the path each `diff --git` line
   names (the new path, `b/…`), so each file renders as its own DiffView. */
export function splitDiff(diff: string): Map<string, string> {
  const out = new Map<string, string>();
  if (!diff) return out;
  const parts = diff.split(/^(?=diff --git )/m);
  for (const part of parts) {
    if (!part.startsWith("diff --git ")) continue;
    const header = part.split("\n", 1)[0];
    const m = /^diff --git a\/(.*) b\/(.*)$/.exec(header);
    const path = m ? m[2] : header.slice("diff --git ".length);
    out.set(path, (out.get(path) || "") + part);
  }
  return out;
}

/* The line under a rewind marker or a result: what the rewind did. */
export function rewindOutcome(result: RewindResult): string {
  const parts: string[] = [];
  if (result.what !== "conversation") {
    const n = (result.restored?.length || 0) + (result.removed?.length || 0);
    parts.push(
      n === 0
        ? "the workspace already matched the checkpoint"
        : `${plural(result.restored?.length || 0, "file")} restored, ${plural(result.removed?.length || 0, "file")} removed`,
    );
  }
  if (result.conversation === "pending")
    parts.push("the agent forgets the messages when its session resumes");
  if (result.conversation === "fresh")
    parts.push(
      "the agent's session could not rewind; the next message starts a new one with the conversation so far as context",
    );
  if (result.withdrawn)
    parts.push(`${plural(result.withdrawn, "queued message")} withdrawn`);
  return parts.join("; ");
}

/* Whether a rewind marker's rewind can be undone now: the service still
   keeps what it removed (`chat.undoRewind` names this marker) and the
   chat is idle. */
export function canUndoRewind(
  entry: Pick<Entry, "id" | "role">,
  chat: Pick<Chat, "status" | "archived" | "undoRewind">,
): boolean {
  return (
    entry.role === "rewind" && chat.undoRewind === entry.id && canRewind(chat)
  );
}

/* Whether an undo may restore the workspace too: the rewind moved the
   code and recorded the workspace as it was first. */
export function undoOffersCode(entry: Pick<Entry, "rewind">): boolean {
  return entry.rewind?.what === "both" && !!entry.rewind.before;
}

/* The undo's hint on the marker: what happens to the workspace. */
export function undoHint(entry: Pick<Entry, "rewind">): string {
  const base = "Undo puts the removed messages back (until the next turn)";
  if (entry.rewind?.what !== "both") return base;
  return undoOffersCode(entry)
    ? base + "; the files can come back too"
    : base + "; the workspace stays as it is";
}

/* The line an undo's result shows: what came back and how the session
   and the workspace followed. */
export function undoOutcome(result: UndoResult): string {
  const parts = [
    `${result.entries === 1 ? "1 entry" : `${result.entries} entries`} restored`,
  ];
  if (result.requeued)
    parts.push(`${plural(result.requeued, "message")} queued again and held`);
  if (result.session === "cancelled")
    parts.push("the agent's session never saw the rewind");
  if (result.session === "resumed")
    parts.push("the agent's session continues where it was");
  if (result.session === "fresh")
    parts.push(
      "the agent's session cannot take the messages back; the next message starts a new one with the conversation so far as context",
    );
  if (result.code === "restored")
    parts.push(
      `the workspace is back as it was before the rewind (${plural(result.restored?.length || 0, "file")} restored, ${plural(result.removed?.length || 0, "file")} removed)`,
    );
  if (result.code === "kept")
    parts.push("the workspace stays as the rewind left it");
  return parts.join("; ");
}

function plural(n: number, unit: string): string {
  return n === 1 ? `1 ${unit}` : `${n} ${unit}s`;
}
